cli/setup/token_fetcher.go

```diff
@@ -0,0 +1,117 @@
+package setup
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+)
+
+func receiveTokenFromUI(c context.Context, serverURL string) (string, error) {
+	port := randomPort()
+
+	tokenReceived := make(chan string)
+
+	srv := &http.Server{Addr: fmt.Sprintf(":%d", port)}
+	srv.Handler = setupRouter(tokenReceived)
+
+	go func() {
+		log.Debug().Msgf("Listening for token response on :%d", port)
+		_ = srv.ListenAndServe()
+	}()
+
+	defer func() {
+		log.Debug().Msg("Shutting down server")
+		_ = srv.Shutdown(c)
+	}()
+
+	err := openBrowser(fmt.Sprintf("%s/cli/auth?port=%d", serverURL, port))
+	if err != nil {
+		return "", err
+	}
+
+	// wait for token to be received or timeout
+	select {
+	case token := <-tokenReceived:
+		return token, nil
+	case <-c.Done():
+		return "", c.Err()
+	case <-time.After(5 * time.Minute):
+		return "", errors.New("timed out waiting for token")
+	}
+}
+
+func setupRouter(tokenReceived chan string) *gin.Engine {
+	gin.SetMode(gin.ReleaseMode)
+	e := gin.New()
+	e.UseRawPath = true
+	e.Use(gin.Recovery())
+
+	e.Use(func(c *gin.Context) {
+		c.Writer.Header().Set("Access-Control-Allow-Origin", "*") // TODO: change to serverURL
```

Comment:
What's missing to add it before we merge it?

Reply:
I think we can actually just ignore it when only listening on localhost.

```suggestion
		c.Writer.Header().Set("Access-Control-Allow-Origin", "*")
```